cpp/src/mlt/util/rle.hpp

```diff
@@ -0,0 +1,189 @@
+#pragma once
+
+#include <mlt/util/buffer_stream.hpp>
+#include <mlt/util/noncopyable.hpp>
+
+#include <algorithm>
+#include <type_traits>
+
+namespace mlt::metadata::stream {
+class StreamMetadata;
+}
+
+namespace mlt::util::decoding::rle {
+
+namespace detail {
+// Borrowed from https://github.com/apache/orc, `/c++/src/ByteRLE.cc`, `ByteRleDecoderImpl::nextInternal`
+// Apache License 2.0
+class ByteRleDecoder : public util::noncopyable {
+public:
+    ByteRleDecoder() = delete;
+    ByteRleDecoder(const std::uint8_t* buffer, size_t length) noexcept
+        : bufferStart(reinterpret_cast<const char*>(buffer)),
+          bufferEnd(bufferStart + length) {}
+    ByteRleDecoder(ByteRleDecoder&&) = delete;
+    ByteRleDecoder& operator=(const ByteRleDecoder&) = delete;
+    ByteRleDecoder& operator=(ByteRleDecoder&&) = delete;
+
+    std::size_t getBufferRemaining() const noexcept { return bufferEnd - bufferStart; }
+
+    void next(std::uint8_t* data,
+              std::uint64_t numValues
+#if SUPPORT_SKIP_NULL
+              ,
+              const char* notNull
+#endif
+              ) noexcept(false) {
+        std::uint64_t position = 0;
+
+#if SUPPORT_SKIP_NULL
+        // skip over null values
+        while (notNull && position < numValues && !notNull[position]) {
+            position += 1;
+        }
+#endif
+
+        while (position < numValues) {
+            // if we are out of values, read more
+            if (remainingValues == 0) {
+                readHeader();
+            }
+            // how many do we read out of this block?
+            size_t count = std::min(static_cast<size_t>(numValues - position), remainingValues);
+            uint64_t consumed = 0;
+            if (repeating) {
+#if SUPPORT_SKIP_NULL
```

Comment:
Is there a good reason to make this configurable?

Reply:
I couldn't tell if it was actually used while porting it over.  I'm 90% sure it isn't now that I've finished most of the decoder branches, so yeah I'll remove it.